include/measurement_kit/nettest.hpp

```diff
@@ -0,0 +1,1922 @@
+
+    /// Logs the "failure.resolver_lookup" event on std::clog.
+    void on_failure_resolver_lookup(const events::FailureResolverLookupEvent &event) override;
+
+    /// Logs the "failure.startup" event on std::clog.
+    void on_failure_startup(const events::FailureStartupEvent &event) override;
+
+    /// Logs the "log" event on std::clog.
+    void on_log(const events::LogEvent &event) override;
+
+    /// Logs the "measurement" event on std::clog.
+    void on_measurement(const events::MeasurementEvent &event) override;
+
+    /// Logs the "status.end" event on std::clog.
+    void on_status_end(const events::StatusEndEvent &event) override;
+
+    /// Logs the "status.geoip_lookup" event on std::clog.
+    void on_status_geoip_lookup(const events::StatusGeoipLookupEvent &event) override;
+
+    /// Logs the "status.progress" event on std::clog.
+    void on_status_progress(const events::StatusProgressEvent &event) override;
+
+    /// Logs the "status.queued" event on std::clog.
+    void on_status_queued(const events::StatusQueuedEvent &event) override;
+
+    /// Logs the "status.measurement_start" event on std::clog.
+    void on_status_measurement_start(const events::StatusMeasurementStartEvent &event) override;
+
+    /// Logs the "status.measurement_submission" event on std::clog.
+    void on_status_measurement_submission(const events::StatusMeasurementSubmissionEvent &event) override;
+
+    /// Logs the "status.measurement_done" event on std::clog.
+    void on_status_measurement_done(const events::StatusMeasurementDoneEvent &event) override;
+
+    /// Logs the "status.report_close" event on std::clog.
+    void on_status_report_close(const events::StatusReportCloseEvent &event) override;
+
+    /// Logs the "status.report_create" event on std::clog.
+    void on_status_report_create(const events::StatusReportCreateEvent &event) override;
+
+    /// Logs the "status.resolver_lookup" event on std::clog.
+    void on_status_resolver_lookup(const events::StatusResolverLookupEvent &event) override;
+
+    /// Logs the "status.started" event on std::clog.
+    void on_status_started(const events::StatusStartedEvent &event) override;
+
+    /// Logs the "status.update.performance" event on std::clog.
+    void on_status_update_performance(const events::StatusUpdatePerformanceEvent &event) override;
+
+    /// Logs the "status.update.websites" event on std::clog.
+    void on_status_update_websites(const events::StatusUpdateWebsitesEvent &event) override;
+
+    /// Logs the "task_terminated" event on std::clog.
+    void on_task_terminated(const events::TaskTerminatedEvent &event) override;
+
+    ~NoisyRouter() noexcept override;
+};
+
+} // namespace routers
+
+/// Manages the lifecycle of a nettest. This API mirrors as closely as
+/// possible to FFI API provided by <measurement_kit/ffi.h>.
+class Nettest {
+  public:
+    // C++ object model
+    // ````````````````
```

Comment:
The object model is such that one can use stack allocation or `new`. I have decided not to do initialisation in the constructor because I don't want to throw (throwing complicates stuff when using SWIG and, say, Golang). As such, there is a specific method one needs to call to "start" the test.

I have made this class explicitly non-copyable and non-movable for safety reasons. If one needs to have unique or shared ownership, the design allows to use `std::make_unique` or `std::make_shared` quite easily.